spec/src/core/semantics.makam

```diff
@@ -0,0 +1,413 @@
+    eval Values format_intro_s64le format_intro_s64le.
+    eval Values format_intro_s64be format_intro_s64be.
+    eval Values (format_intro_array ElemType LenTerm) (format_intro_array ElemType' LenValue) :-
+        eval Values ElemType ElemType',
+        eval Values LenTerm LenValue.
+    eval Values (format_intro_pair FirstType SecondType) (format_intro_pair FirstType' (closure Values SecondType)) :-
+        eval Values FirstType FirstType'.
+    eval Values (format_elim Term) Value' :-
+        eval Values Term Value,
+        format_elim Value Value'.
+
+    % Closure operations
+
+    apply (closure Values Term) InputValue Value :-
+        eval (InputValue :: Values) Term Value.
+
+    apply_param Closure Length Value :-
+        apply Closure (neutral (local Length)) Value.
+
+    % Eliminations
+
+    function_elim (neutral Neutral) InputValue (neutral (function_elim Neutral InputValue)).
+    function_elim (function_intro Closure) InputValue OutputValue :-
+        apply Closure InputValue OutputValue.
+
+    pair_elim_first (neutral Neutral) (neutral (pair_elim_first Neutral)).
+    pair_elim_first (pair_intro FirstValue _) FirstValue.
+
+    pair_elim_second (neutral Neutral) (neutral (pair_elim_second Neutral)).
+    pair_elim_second (pair_intro _ SecondValue) SecondValue.
+
+    % TODO: stage_elim
+
+    %  bool_elim (neutral Neutral) TrueValue FalseValue (neutral (bool_elim Neutral TrueValue FalseValue)).
+    %  bool_elim (bool_intro true) TrueValue _ TrueValue.
+    %  bool_elim (bool_intro false) _ FalseValue FalseValue.
+
+    % TODO: array_elim
+
+    format_elim (neutral Neutral) (neutral (format_elim Neutral)).
+    format_elim format_intro_void void_type.
+    format_elim format_intro_unit unit_type.
+    format_elim format_intro_u8 int_type.
+    format_elim format_intro_u16le int_type.
+    format_elim format_intro_u16be int_type.
+    format_elim format_intro_u32le int_type.
+    format_elim format_intro_u32be int_type.
+    format_elim format_intro_u64le int_type.
+    format_elim format_intro_u64be int_type.
+    format_elim format_intro_s8 int_type.
+    format_elim format_intro_s16le int_type.
+    format_elim format_intro_s16be int_type.
+    format_elim format_intro_s32le int_type.
+    format_elim format_intro_s32be int_type.
+    format_elim format_intro_s64le int_type.
+    format_elim format_intro_s64be int_type.
+    format_elim (format_intro_array Elem Len) (array_type Elem' Len) :-
+        format_elim Elem Elem'.
+    format_elim (format_intro_pair First (closure Values Term)) (pair_type First' (closure Values (format_elim Term))) :-
+        format_elim First First'.
+
+
+    % Readback of values into terms in normal form
+
+    readback : int -> neutral -> term -> prop.
+    readback : int -> value -> term -> prop.
+
+    % FIXME: Makam's type unification breaks hear for some reason - it
+    % thinks that we want to definine `int -> value -> term -> prop.`
```

Comment:
The issue here is that both `readback` and `local` are overloaded and constant resolution fails as a result. Type-based overloading in Makam does not support arbitrary backtracking, but has a few heuristics that unfortunately are not enough here. One option is to avoid defining `readback : int -> value -> term -> prop` until later, if you don't need to mix the rules of the two predicates.

Reply:
Ahhh right! Didn't realise it didn't do backtracking. 

Yea I can't really re-order because they are mutually recursive. This is where I kind of wish for languages that used names for 'presentation', but had some kind of unique identifier under the hood. So inputting an identifier would be more like a 'search-disambiguate' operation in your editor. Anyway, that's straying a bit off topic here! 😅